### DISCUSSION TITLE
Branch filter is slow when there are many branches

### DISCUSSION THREAD

**Discussion by ebbsenghargor** [gitextensions/gitextensions#9851] [Q&A]
## Current behaviour
When using the branch filter in a repo with many remote branches, branch filtering is extremely slow.

## Expected behaviour
Branch filtering should never be slow. As a user I'd expect that there was a delay after last input before starting to filter.

## Steps to reproduce
1. Open a repo with many remote branches (in my case ~7k)
2. Set branch filtering to look for both local and remote branches
3. Start typing something into the branch filter

Note: The search field in the branch list does not have this issue.

## Screenshots

I don't have a video of the issue since it would be a bit tricky to record something useful without disclosing sensitive information.

For clarity, this screenshot shows the field I'm talking about:
![image](https://user-images.githubusercontent.com/47830737/66269734-8e0ebd80-e84c-11e9-9229-24377076b1e5.png)

## Did this work in previous version of GitExtensions
This is not a recent regression, but I'm not sure if there's a previous version that worked better.

## Environment

- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4010.0
- DPI 96dpi (no scaling)

**Comment by big-jon** [comment]
Same for my repo with ~5000 branches. Each character typed into the branch filter is incredibly slow, roughly half a second per character.

This is still an issue in the following environment:
- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.32.0.windows.2
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

**Comment by vbjay** [comment]
Remove obsolete branches plugin would prob help the 5000 branch issue.

**Comment by RussKie** [comment]
These are "quick" filters. With so many branches use the Advanced filter
(under the View menu).

**Comment by big-jon** [comment]
I will have to look into the obsolete branches plugin further, but currently it is only showing 34 obsolete branches (including remotes). Likely due to my company workflow of rebase/squash before committing to mainline. I have no doubt we could remove 4k+ of the branches.

I'm not sure if the issue is clear. I only want to view a single branch, like: "personal/user_name/issue_id". However, after each character is typed, GitExtensions will search for branches matching what is typed so far. But the search is slow, so typing out "personal/user_name" is painfully slow. My request is to modify gitExtensions to only perform the search when no character has been typed for some amount of time (500ms? 250ms?). Then I will be able to type "personal/user_name" quickly, then select my current issue branch from the dopdown.

**Comment by RussKie** [comment]
Check out https://github.com/not-an-aardvark/git-delete-squashed
